File: LeetCode/19. Remove Nth Node From End of List/elixir/solution.exs
# Definition for singly-linked list.
#
# defmodule ListNode do
#   @type t :: %__MODULE__{
#           val: integer,
#           next: ListNode.t() | nil
#         }
#   defstruct val: 0, next: nil
# end

defmodule Solution do
  @spec remove_nth_from_end(head :: ListNode.t() | nil, n :: integer) :: ListNode.t() | nil
  def remove_nth_from_end(%{next: nil}, _n) do
    nil
  end

  def remove_nth_from_end(head, n) do
    head
    |> to_value_list()
    |> List.delete_at(n - 1)
    |> to_list_node_struct()
  end

  defp to_value_list(node, acc \\ []) do
    case node do
      %{next: nil} -> [node.val | acc]
      %{next: next} -> to_value_list(next, [node.val | acc])
    end
  end

  defp to_list_node_struct(list, acc_node \\ nil)

  defp to_list_node_struct([], acc_node) do
    acc_node
  end

  defp to_list_node_struct([h | t], acc_node) do
    to_list_node_struct(t, %ListNode{val: h, next: acc_node})
  end
end

File: LeetCode/19. Remove Nth Node From End of List/java/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int length = length(head);
        int removeIndex = length - n;
        if (removeIndex == 0) {
            return head.next;
        }
        ListNode node = head;
        for (int i = 0; node != null && i < length; i++) {
            if (i == removeIndex - 1) {
                node.next = node.next.next;
            }
            node = node.next;
        }
        return head;
    }

    private int length(ListNode head) {
        return length(head, 1);
    }

    private int length(ListNode node, int acc) {
        if (node.next == null) {
            return acc;
        }
        return length(node.next, acc + 1);
    }
}
